### PR TITLE
Explicit Message Padding for OnionRealm0HopData

### LINKED ISSUE #467: Use zero-sized field for message padding
Could you use something like `PhantomData<[u8; 12]>` (or our own custom `Padding` type) for this? It could allow for using zero bytes yet still have a field for reading/writing the data. Then we'd make reading the padding explicit rather than doing so while reading `outgoing_cltv_value`.

_Originally posted by @jkczyz in https://github.com/rust-bitcoin/rust-lightning/pull/434_

### PR DESCRIPTION
**Pull request by varunsrin**
Fixes #467

### CODE REVIEW

File: lightning/src/ln/msgs.rs [hunk 1]
```diff
@@ -610,7 +611,7 @@ pub(crate) struct OnionRealm0HopData {
 	pub(crate) short_channel_id: u64,
 	pub(crate) amt_to_forward: u64,
 	pub(crate) outgoing_cltv_value: u32,
-	// 12 bytes of 0-padding
+	pub(crate) padding: PhantomData<[u8; 12]>,
```
Comment (jkczyz) [inline lightning/src/ln/msgs.rs]: It's too bad RFC 2000 (https://github.com/rust-lang/rust/issues/44580) isn't in the stable Rust release yet. Then we could type alias this as:

```rust
pub(crate) type BytePadding<const N: usize> = ::std::marker::PhantomData<[u8; N]>;
```

So this should be fine for now.

File: lightning/src/util/ser.rs [hunk 2]
```diff
@@ -392,6 +394,20 @@ impl Writeable for PaymentHash {
...
+impl<R: Read> Readable<R> for PhantomData<[u8; 12]> {
+	fn read(r: &mut R) -> Result<Self, DecodeError> {
+		let _buf: [u8; 12] = Readable::read(r)?; 
+		Ok(PhantomData)
+	}
+}
+
+impl Writeable for PhantomData<[u8; 12]> {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		w.write_all(&[0;12])?;
```
Comment (jkczyz) [inline lightning/src/util/ser.rs]: Nit: Space between `;` and `12`.

Comment (varunsrin) [reply lightning/src/util/ser.rs]: fixed

File: lightning/src/ln/msgs.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs]: We should probably do something like this anyway without generic constants. PhantomData has a somewhat specific meaning in Rust (that we have a template type that we're not going to instantiate). We probably just want a new type in ser.rs like (but with better names):
```

pub(crate) struct PaddingData<T: Default + std::convert::AsRef<[u8]>> {
t: std::marker::PhantomData<T>,
}
impl<T: Default + std::convert::AsRef<[u8]>> Writeable for PaddingData<T> {
	#[inline]
	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
        let nt: T = Default::default();
        w.write_all(nt.as_ref())
	}
}
```

**Comment by jkczyz** [comment]
@varunsrin #434 has been merged, so you should be able to rebase now.

**Comment by varunsrin** [comment]
perfect, ill get to this by early next week

**Comment by TheBlueMatt** [comment]
Closing as abandoned. If you want to pick this back up just respond here and we can open it again :)

**Pull request closed without merging**